ably/realtime_client.go

```diff
@@ -5,20 +5,30 @@ import (
 	"time"
 )
 
+// The RealtimeV12 libraries establish and maintain a persistent connection
+// to Ably enabling extremely low latency broadcasting of messages and presence
+// state.
+type RealtimeV12 = RealtimeClient
```

Comment:
Given this is being merged into the `integration/1.2` branch, I wonder whether we should just actively break the API in this branch to be compatible with the 1.2 spec, and then one day we just switch, make `integration/1.2` be `master` and then maintain a deprecated `release-1.1` branch with security fixes.

This would make the code easier to maintain (no need for multiple implementations of the same things in the same codebase but with `V12` appended to their names), and makes upgrading to `1.2` a one step process (i.e. just update to the `1.2` branch in `go.mod` and continue to use `ably.NewRealtimeClient` but with newer features, rather than using `ably.NewRealtimeClientV12` for a period of time and then in the future needing to go back and change it back to `ably.NewRealtimeClient`).

WDYT?

/cc @QuintinWillison how does this work in other repositories?

Reply:
I've not seen anything similar elsewhere so, as far as I know, this is new territory for Ably... though others have much more experience than me in how we do things here.

I'm not sure I entirely understand the problem in hand, to be honest. And, on top of that, I'm looking at the names `NewRealtimeClientV12` and `NewRealtimeClient` and wondering why they've even got a `New` prefix in the first place!? Those names already make it feel like more than a warm start, if that makes sense!

Reply:
I really don't think we should be annotating type or function names with the version. If we're being forced to make a breaking change because what we had was broken in a way that prevents us being compatible, then we have to have a step where we force an upgrade and require the user to make changes.

> I'm looking at the names NewRealtimeClientV12 and NewRealtimeClient and wondering why they've even got a New prefix in the first place!?

That's how factory functions are conventionally named in go.

Reply:
@QuintinWillison 

> I'm not sure I entirely understand the problem in hand

So the main issue is maintaining backwards compatibility with people who may need to upgrade the ably-go dependency for reasons other than needing the new API we are shipping.

For example, if we push a security fix, people should be able to upgrade easily to take advantage of that fix without needing to also refactor their code because we also broke the API.

The way people upgrade in Go is typically by just pulling in master with `go get -u`, although recently Go got native module support which makes it easier to pull in changes from a specific branch (e.g. `go get -u github.com/ably/ably-go@release-1.1`).

In terms of us maintaining this backwards compatibility, we have two options:

1) have a single git branch that supports both the new and the old API (which is what @tcard is doing in this PR so far) and expect users to specify in their code which version they are using (e.g. `ably.NewRealtimeClientV12()`)

2) have a branch per potentially incompatible version (e.g. `release-1.1` and `release-1.2`) and expect users to specify in their dependency file (i.e. `go.mod`) which version of the API they are using, and then use that in their code

Personally, I think 2) is easier to maintain, and makes a clear difference between upgrading for a security fix and expecting no breaking changes (i.e. pulling in the latest commits from the `release-1.1` branch) vs upgrading for a new feature and expecting to change your code accordingly (i.e. switching to the `release-1.2` branch, or `master`).

I also think we should have some form of policy we follow across repositories for this.

Reply:
> I really don't think we should be annotating type or function names with the version.

Just to clarify, this is temporary and won't be there in any public-facing release. See [this comment](https://github.com/ably/ably-go/pull/144#discussion_r404234774) for rationale.

Reply:
About backwards compatibility, basically what @lmars says is ultimately the plan here, except following Go modules format for version tags. And, of course, `master` will point to the newest release.

Reply:
Ultimately we decided on a breaking/replacing approach rather than appending for implementing the new API. Removed the `V12` suffixes at db867ccad5102c4af03796781ff973ad5b6fcb7f, 7d387f5882c63cf85d5376c223d2f98d7d523a0a.